### PR TITLE
ARROW-1096: [C++] CreateFileMapping maximum size calculation issue

### PR DESCRIPTION
**Pull request by maxhora**

### CODE REVIEW

File: cpp/src/arrow/io/mman.h [hunk 1]
```diff
@@ -93,13 +93,10 @@ static void* mmap(void* addr, size_t len, int prot, int flags, int fildes, off_t
   const DWORD protect = __map_mmap_prot_page(prot);
   const DWORD desiredAccess = __map_mmap_prot_file(prot);
 
-  const off_t maxSize = off + (off_t)len;
```
Comment (MaxRis) [inline cpp/src/arrow/io/mman.h]: MSVC defines `off_t` as
```
typedef long _off_t; // file offset value
```
`long` type width is 32 bits and it's not enough to keep 64 bits of size_t(unsigned long long).
File size over 4Gb needs more than 32 bits, that caused the issue.

Comment (MaxRis) [inline cpp/src/arrow/io/mman.h]: MSVC defines `off_t` as
```
typedef long _off_t; // file offset value
```
msvc `long` type width is 32 bits and it's not enough to keep 64 bits of size_t(unsigned long long).
File size over 4Gb needs more than 32 bits, that caused the issue.

**Comment by MaxRis** [comment]
Going to add some unit test for this

**Comment by wesm** [comment]
did you close the PR by accident? ok to leave it open

**Comment by MaxRis** [comment]
@wesm , I'm adding unit test for it, will reopen again

**Comment by wesm** [comment]
How are you planning to test? creating a 4GB file on Appveyor may not be possible

**Comment by MaxRis** [comment]
@wesm , yes, I've added generation of ~ 4,6 GB file ... on Appveyor it's passed fine, I've also verified the size of the file manually locally, so the size of file is correct

**Pull request closed without merging**